src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -1013,11 +1014,27 @@ fn preprocess_link(
     // | has backtick |    never ignore    |    never ignore   |
     // | no backtick  | ignore if url-like |    never ignore   |
     // |-------------------------------------------------------|
-    let ignore_urllike =
-        can_be_url || (ori_link.kind == LinkType::ShortcutUnknown && !ori_link.link.contains('`'));
+    let ignore_urllike = can_be_url || (is_shortcut_style && !ori_link.link.contains('`'));
     if ignore_urllike && should_ignore_link(path_str) {
         return None;
     }
+    // Ignore GitHub-flavored Markdown (GFM) admonitions, such as [!NOTE] and [!IMPORTANT]
+    //
+    // rustdoc does not support GFM,
+    // however it is a common pattern to add `#[doc = include_str!("../README.md")]` to the root of a crate,
+    // so we want to at least accept GFM, even if it doesn't render perfectly.
+    //
+    // We make sure to allow `[!]` as a link to the never type by checking length.
+    //
+    // This will never cause us to ignore something that is a valid intra-doc link,
+    // as (with the exception of `[!]`), intra-doc links can never start with `!`.
```

Comment:
```suggestion
    // If we have an intra-doc link starting with `!` (which isn't `[!]` because this is the never type), we ignore it
    // as it is never valid.
    //
    // The case is common enough because of cases like `#[doc = include_str!("../README.md")]` which often
    // uses GitHub-flavored Markdown (GFM) admonitions, such as `[!NOTE]`.
```